auth/services/oauth/holder.go

```diff
@@ -0,0 +1,139 @@
+ *
+ */
+
+package oauth
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/auth/client/iam"
+	"github.com/nuts-foundation/nuts-node/auth/oauth"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/vcr/holder"
+	"github.com/nuts-foundation/nuts-node/vcr/pe"
+	"github.com/nuts-foundation/nuts-node/vcr/signature/proof"
+	"time"
+)
+
+var _ Holder = (*HolderService)(nil)
+
+var ErrNoCredentials = errors.New("no matching credentials")
+
+type HolderService struct {
+	strictMode        bool
+	httpClientTimeout time.Duration
+	httpClientTLS     *tls.Config
+	wallet            holder.Wallet
+}
+
+// NewHolder returns an implementation of Holder
+func NewHolder(wallet holder.Wallet, strictMode bool, httpClientTimeout time.Duration, httpClientTLS *tls.Config) *HolderService {
+	return &HolderService{
+		wallet:            wallet,
+		strictMode:        strictMode,
+		httpClientTimeout: httpClientTimeout,
+		httpClientTLS:     httpClientTLS,
+	}
+}
+
+func (v *HolderService) BuildPresentation(ctx context.Context, walletDID did.DID, presentationDefinition pe.PresentationDefinition, acceptedFormats map[string]map[string][]string, nonce string) (*vc.VerifiablePresentation, *pe.PresentationSubmission, error) {
+	// get VCs from own wallet
+	credentials, err := v.wallet.List(ctx, walletDID)
+	if err != nil {
+		return nil, nil, errors.New("failed to retrieve wallet credentials")
+	}
+
+	expires := time.Now().Add(time.Minute * 15) //todo
+	// build VP
+	submissionBuilder := presentationDefinition.PresentationSubmissionBuilder()
+	submissionBuilder.AddWallet(walletDID, credentials)
+	format := pe.ChooseVPFormat(acceptedFormats)
+	presentationSubmission, signInstructions, err := submissionBuilder.Build(format)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to build presentation submission: %w", err)
+	}
+	if signInstructions.Empty() {
+		return nil, nil, ErrNoCredentials
+	}
+
+	// todo: support multiple wallets
+	vp, err := v.wallet.BuildPresentation(ctx, signInstructions[0].VerifiableCredentials, holder.PresentationOptions{
+		Format: format,
+		ProofOptions: proof.ProofOptions{
+			Created:   time.Now(),
+			Challenge: &nonce,
+			Expires:   &expires,
+		},
+	}, &walletDID, false)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create verifiable presentation: %w", err)
+	}
+	return vp, &presentationSubmission, nil
+}
+
+func (v *HolderService) ClientMetadata(ctx context.Context, endpoint string) (*oauth.OAuthClientMetadata, error) {
+	iamClient := iam.NewHTTPClient(v.strictMode, v.httpClientTimeout, v.httpClientTLS)
+	// The verifier has become the client
+
+	metadata, err := iamClient.ClientMetadata(ctx, endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve remote OAuth Authorization Server metadata: %w", err)
```

Comment:
should we include the URL for debuggin?

Reply:
I believe most errors from the Go core library already include it.